File: 2398/main.go
package main

import "fmt"

func main() {
	fmt.Println(maximumRobots([]int{1,2,3,4,5}, []int{1,2,3,4,5}, 25))
}

func maximumRobots(chargeTimes []int, runningCosts []int, budget int64) int {
	res, n := 0, len(chargeTimes)
	runningCostSum := int64(0)
	var q []int
	for i, j := 0, 0; i < n; i++ {
		runningCostSum += int64(runningCosts[i])
		for len(q) > 0 && chargeTimes[q[len(q)-1]] <= chargeTimes[i] {
			q = q[:len(q)-1]
		}
		q = append(q, i)
		for j <= i && int64(i - j + 1) * runningCostSum + int64(chargeTimes[q[0]]) > budget {
			if len(q) > 0 &&  q[0] == j {
				q = q[1:]
			}
			runningCostSum -= int64(runningCosts[j])
			j++
		}
		res = max(res, i -j + 1)
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 2398/main.rs
use std::collections::VecDeque;

struct Solution {}

impl Solution {
    pub fn maximum_robots(charge_times: Vec<i32>, running_costs:Vec<i32>, budget: i64) -> i32 {

        let mut res = 0;
        let n = charge_times.len();
        let mut running_const_sum: i64 = 0;
        let mut q: VecDeque<usize> = VecDeque::new();
        let mut j = 0;
        for i in 0..n {
            running_const_sum += running_costs[i] as i64;
            while !q.is_empty() && charge_times[*q.back().unwrap()] <= charge_times[i] {
                q.pop_back();
            }
            q.push_back(i);
            while j <= i && (i - j + 1) as i64 * running_const_sum + charge_times[*q.front().unwrap()] as i64 > budget {
                if !q.is_empty() &&  *q.front().unwrap() == j {
                    q.pop_front();
                }
                running_const_sum -= running_costs[j] as i64;
                j += 1;
            }
            res = res.max(i - j + 1);
        }
        return res as i32
    }
}

fn main() {
    println!("{}", Solution::maximum_robots(vec![1,2,3,4,5], vec![1,2,3,4,5], 25))
}